File: hack/url-scheme-parity/Parity.java
// Parity reproduces how Spark decides whether to fetch a dependency value, using
// the same java.net.URI parsing Spark uses, so it can be diffed against parity.go
// (which reproduces the operator webhook's Go net/url decision).
//
// Spark path modelled (Spark 3.x/4.x, unchanged in sfdc-spark):
//   - org.apache.spark.util.SparkFileUtils.resolveURI: new URI(path); if getScheme()
//     is null it becomes a local file:// path; a URISyntaxException also falls back
//     to a local File.
//   - org.apache.spark.util.Utils.downloadFile: switches on
//     Option(uri.getScheme).getOrElse("file") -> "spark" (RPC),
//     "http"|"https"|"ftp" (URL.openConnection), "file" (local copy),
//     otherwise Hadoop FileSystem (fetch). The http/https/ftp match is
//     case-sensitive against the original-case scheme.
//
// Run directly (JDK 11+, no separate compile step):
//   java Parity.java > /tmp/java.txt
//
// The security-relevant question: is there any input classified FETCH-* here that
// parity.go classifies ALLOW? Diff the two outputs; there should be none.
import java.net.URI;
import java.util.Locale;

public class Parity {

    // INPUTS must stay byte-identical (and same order) to inputs in parity.go.
    static final String[] INPUTS = new String[] {
        "http://evil/x",
        "HTTP://evil/x",
        "https://evil/x",
        "ftp://evil/x",
        "s3a://bucket/x",
        "hdfs://nn/x",
        "gs://bucket/x",
        "gopher+ssh://x",
        "http:evil/x",       // opaque, no //
        "s3a:/onlyone",      // single slash
        "file:///opt/a.jar",
        "local:///opt/a.jar",
        "file://host/opt/a.jar",
        "local://host/opt/a.jar",
        "/abs/path",
        "./rel/path",
        "c:/windows/path",   // Windows drive letter reads as scheme "c"
        "mailto:a@b",
        "//host/path",       // scheme-relative
        " http://evil/x",    // leading space
        "http\t://evil/x",   // embedded tab (control char)
        "java\nscript:x",    // embedded newline
        "ht!tp://x",         // illegal scheme char
        "1http://x",         // scheme cannot start with a digit
        "a\\b://x",          // backslash
        "http ://x",         // space before colon
    };

    // decide mirrors resolveURI + downloadFile: what Spark would do with the value.
    // Returns "VERDICT\tSCHEME".
    static String decide(String value) {
        String scheme;
        try {
            scheme = new URI(value).getScheme();
        } catch (Exception e) {
            // resolveURI catches URISyntaxException and treats the value as a local File.
            return "LOCAL\t<unparseable>";
        }
        if (scheme == null) {
            return "LOCAL\t<none>";       // resolveURI -> file:// local path
        }
        // downloadFile matches http/https/ftp case-sensitively on the raw scheme.
        switch (scheme) {
            case "spark":
                return "RPC\t" + scheme;
            case "http":
            case "https":
            case "ftp":
                return "FETCH-HTTP\t" + scheme;
            case "file":
                return "LOCAL\t" + scheme;
            case "local":
                // "local://" is Spark's in-container scheme: the resource is expected
                // to already exist in the driver/executor image and is never fetched by
                // the submitter. KubernetesUtils treats it as a non-local-dependency that
                // is left as-is, so the operator does not dereference it. Not a vector.
                return "LOCAL\t" + scheme;
            default:
                // Everything else (s3a, hdfs, gs, HTTP, c, mailto, ...) hits the
                // Hadoop FileSystem branch: getHadoopFileSystem(uri) + fetch. Some
                // of these fail later for lack of a filesystem impl, but the fetch
                // is attempted with the submitter's (operator's) credentials.
                return "FETCH-HADOOP\t" + scheme;
        }
    }

    static String show(String s) {
        return s.replace("\\", "\\\\").replace("\t", "\\t").replace("\n", "\\n");
    }

    public static void main(String[] args) {
        System.out.printf("%-13s %-12s %s%n", "VERDICT", "SCHEME", "INPUT");
        for (String in : INPUTS) {
            String[] parts = decide(in).split("\t", 2);
            String verdict = parts[0];
            String scheme = parts.length > 1 ? parts[1] : "";
            // Lower-case display of scheme to line up with Go's lowercased output;
            // the switch above already used the raw case for the fetch decision.
            System.out.printf("%-13s %-12s %s%n", verdict, scheme.toLowerCase(Locale.ROOT), show(in));
        }
    }
}

File: hack/url-scheme-parity/parity.go
// Command parity reproduces the webhook's URL-scheme decision (internal/webhook
// checkURLScheme, with an empty allow-list) using Go's net/url, so it can be
// diffed against Parity.java which reproduces how Spark decides whether to fetch
// a value (java.net.URI.getScheme, as used by SparkFileUtils.resolveURI and
// Utils.downloadFile).
//
// The two programs share the same INPUTS list. The security-relevant question
// is one-directional: is there any input the webhook ALLOWs but Spark would
// FETCH remotely? Run both and diff:
//
//	go run parity.go            > /tmp/go.txt
//	java Parity.java            > /tmp/java.txt
//	paste -d' ' /tmp/go.txt /tmp/java.txt   # eyeball, or diff the VERDICT columns
//
// A safe result is: every row where Java prints FETCH-* has Go printing REJECT.
package main

import (
	"fmt"
	"net/url"
	"strings"
)

// inputs is the shared adversarial corpus. Keep byte-identical to INPUTS in
// Parity.java (same order) so the two outputs line up row for row.
var inputs = []string{
	"http://evil/x",
	"HTTP://evil/x",
	"https://evil/x",
	"ftp://evil/x",
	"s3a://bucket/x",
	"hdfs://nn/x",
	"gs://bucket/x",
	"gopher+ssh://x",
	"http:evil/x",     // opaque, no //
	"s3a:/onlyone",    // single slash
	"file:///opt/a.jar",
	"local:///opt/a.jar",
	"file://host/opt/a.jar",
	"local://host/opt/a.jar",
	"/abs/path",
	"./rel/path",
	"c:/windows/path", // Windows drive letter reads as scheme "c"
	"mailto:a@b",
	"//host/path",       // scheme-relative
	" http://evil/x",    // leading space
	"http\t://evil/x",   // embedded tab (control char)
	"java\nscript:x",    // embedded newline
	"ht!tp://x",         // illegal scheme char
	"1http://x",         // scheme cannot start with a digit
	"a\\b://x",          // backslash
	"http ://x",         // space before colon
}

// alwaysAllowed mirrors internal/webhook.alwaysAllowedURLSchemes. All local forms
// must also have an empty authority; see decide.
var alwaysAllowed = map[string]struct{}{"": {}, "file": {}, "local": {}}

// decide reproduces checkURLScheme with an empty operator allow-list: only the
// always-allowed local schemes with no authority pass; anything else, an authority-bearing
// local form, or a value net/url cannot parse (fail closed) is rejected.
func decide(value string) (verdict, scheme, detail string) {
	value = strings.TrimSpace(value)
	if value == "" {
		return "ALLOW", "", "empty"
	}
	u, err := url.Parse(value)
	if err != nil {
		return "REJECT", "", "parse error (fail closed): " + err.Error()
	}
	scheme = strings.ToLower(u.Scheme)
	if _, ok := alwaysAllowed[scheme]; ok {
		if u.Host != "" {
			return "REJECT", scheme, "authority " + u.Host + " not local"
		}
		return "ALLOW", scheme, "always-allowed local scheme"
	}
	return "REJECT", scheme, "scheme not in allow-list"
}

func main() {
	fmt.Printf("%-8s %-12s %s\n", "VERDICT", "SCHEME", "INPUT")
	for _, in := range inputs {
		verdict, scheme, _ := decide(in)
		fmt.Printf("%-8s %-12s %s\n", verdict, scheme, quote(in))
	}
}

// quote renders control chars visibly so rows stay single-line and line up with
// the Java output.
func quote(s string) string {
	r := strings.NewReplacer("\t", "\\t", "\n", "\\n", "\\", "\\\\")
	return r.Replace(s)
}
